client/asset/btc/spv_peer_manager.go

```diff
@@ -0,0 +1,236 @@
+
+package btc
+
+import (
+	"encoding/binary"
+	"fmt"
+	"net"
+	"path/filepath"
+	"sync"
+
+	"decred.org/dcrdex/client/asset"
+	"decred.org/dcrdex/client/asset/kvdb"
+	"decred.org/dcrdex/dex"
+)
+
+type peerSource uint16
+
+const (
+	added peerSource = iota
+	defaultPeer
+)
+
+func (s peerSource) toAssetPeerSource() asset.PeerSource {
+	switch s {
+	case added:
+		return asset.UserAdded
+	case defaultPeer:
+		return asset.WalletDefault
+	}
+	return asset.Discovered
+}
+
+func (s peerSource) MarshalBinary() ([]byte, error) {
+	b := make([]byte, 16)
+	binary.BigEndian.PutUint16(b, uint16(s))
+	return b, nil
+}
+
+type walletPeer struct {
+	source peerSource
+	// resolvedNames is used to avoid adding duplicate peers
+	resolvedNames []string
+}
+
+// PeerManagerChainService are the functions needed for an SPVPeerManager
+// to communicate with a chain service.
+type PeerManagerChainService interface {
+	AddPeer(addr string) error
+	RemoveNodeByAddr(addr string) error
+	Peers() []SPVPeer
+}
+
+// SPVPeerManager implements peer management functionality for all bitcoin
+// clone SPV wallets.
+type SPVPeerManager struct {
+	cs PeerManagerChainService
+
+	peersMtx sync.RWMutex
+	peers    map[string]*walletPeer
+	peersDB  kvdb.KeyValueDB
+
+	defaultPeers []string
+
+	log dex.Logger
+}
+
+// Peers returns the list of peers that the wallet is connected to. It also
+// returns the peers that the user added that the wallet may not currently
+// be connected to.
+func (w *SPVPeerManager) Peers() ([]*asset.WalletPeer, error) {
+	w.peersMtx.RLock()
+	defer w.peersMtx.RUnlock()
+
+	peers := w.cs.Peers()
+	peersMap := make(map[string]*asset.WalletPeer, len(peers))
+
+	for _, peer := range peers {
+		walletPeer := &asset.WalletPeer{
+			Host:      peer.Addr(),
+			Connected: true,
+			Source:    asset.Discovered,
+		}
+
+		if peer, found := w.peers[peer.Addr()]; found {
+			walletPeer.Source = peer.source.toAssetPeerSource()
+		}
+
+		peersMap[walletPeer.Host] = walletPeer
+	}
+
+	// Also return the peers that the user has added but are not connected
+	for host, peer := range w.peers {
+		if _, found := peersMap[host]; !found {
+			walletPeer := &asset.WalletPeer{
+				Host:   host,
+				Source: peer.source.toAssetPeerSource(),
+			}
+			peersMap[host] = walletPeer
+		}
+	}
+
+	walletPeers := make([]*asset.WalletPeer, 0, len(peersMap))
+	for _, peer := range peersMap {
+		walletPeers = append(walletPeers, peer)
+	}
+
+	return walletPeers, nil
+}
+
+// peerWithResolvedNames checks if there is a peer which has a resolved name
+// which matches one of the resolvedNames passed into the function.
+func (w *SPVPeerManager) peerWithResolvedNames(resolvedNames []string) (string, bool) {
+	for name, peer := range w.peers {
+		for _, peerResolvedName := range peer.resolvedNames {
+			for _, newPeerResolvedName := range resolvedNames {
+				if peerResolvedName == newPeerResolvedName {
+					return name, true
+				}
+			}
+		}
+	}
+	return "", false
+}
```

Comment:
I think address resolution is unnecessary once you do @JoeGruffins suggestion. Should be able to clean things up a little bit.

Reply:
I still have the address resolution.. I wrote the reason in the comment here:
https://github.com/decred/dcrdex/blob/52d2b69fd4128c3d69206224be7c02e78e9c0cba/client/asset/btc/spv_peer_manager.go#L103

Another option would be to update the neutrino code to return the resolved address it is using from `ConnectNode`.